### PR TITLE
One program identifier

### LINKED ISSUE #580: kernel ids - the one true program identifier
bpfd currently allocates a UUID for every loaded program.
this serves a few purposes:

1. It gives us a sentinel value that we can use for storing things on the filesystem
2. It helps us stay in sync when bpfd is driven by an orchestrator
3. Our UUIDs *should* be unique across the cluster

However, the kernel has it's own program/link/map identifiers.
- These are a u32, you provided by the kernel at load/create time.
- They are locally significant

We should preferably have only one unique means of identifying programs when using bpfd/bpfctl.
This should be the kernel ID for consistency with other tools, like bpftool.

Therefore, any UUID from an orchestrator should get stored in the program metadata instead.
Map pin paths will no longer be easily used from K8s - this now needs to be fixed in the design of #522

### PR DESCRIPTION
**Pull request by astoycos**
closes and pull's from #656 (thanks @Billy99 For fixing up the examples 🥳  I'll fix up the commit so it says owned by you before merging)

Fixes #580

### CODE REVIEW

**Comment by astoycos** [comment]
There's still some small spelling and nit stuff in comments but otherwise this should be read for review

**Comment by astoycos** [comment]
Added an extra commit here that fixes filtering by metadata + adds an integration test to ensure it's working.

File: bpfd/src/bpf.rs [hunk 1]
```diff
@@ -66,64 +65,67 @@ impl ProgramMap {
...
-            p.kind() == *program_type && p.if_index() == *if_index && p.direction() == *direction
+    ) -> impl Iterator<Item = &'a mut Program> {
+        self.programs.iter_mut().filter_map(|(_, p)| {
+            if p.kind() == *program_type && p.if_index() == *if_index && p.direction() == *direction
+            {
+                Some(p)
+            } else {
+                None
+            }
         })
```
Comment (dave-tucker) [inline bpfd/src/bpf.rs:97]: Use `values_mut()`

```suggestion
        self.programs.values_mut().filter(|p| {
```

**Review by dave-tucker — CHANGES_REQUESTED** [review]

**Review by dave-tucker — CHANGES_REQUESTED** [review]
A few nits but overall looking good. I haven't reviewed the Go code though.

**Review by dave-tucker — CHANGES_REQUESTED** [review]
A few nits but overall looking good. I haven't reviewed the Go code though.

File: bpfd/src/bpf.rs [hunk 2]
```diff
@@ -66,64 +65,67 @@ impl ProgramMap {
...
                     if v.if_index() == if_index && v.direction() == direction {
-                        Some((k, v))
+                        Some(v)
                     } else {
                         None
                     }
                 } else {
                     None
                 }
             })
```
[the remark this answers is not in the document]
Comment (astoycos) [reply bpfd/src/bpf.rs:122]: Arg good catch thanks!

**Review by dave-tucker — APPROVED** [review]

File: bpfd/src/bpf.rs [hunk 3]
```diff
@@ -875,66 +868,71 @@ impl BpfManager {
...
-            let program = self
-                .programs
-                .get_mut(&id)
-                .expect("Program should be loaded");
-
-            let map = BpfMap { used_by: vec![id] };
-
-            self.maps.insert(id, map);
+        match map_owner_id {
+            Some(id) => {
```
Comment (Billy99) [inline bpfd/src/bpf.rs:880]:
```suggestion
            Some(owner_id) => {
```
`id` already exists and needs to be used in this case statement. Change `id` to something else in this case statement.

File: bpfd/src/bpf.rs [hunk 4]
```diff
@@ -875,66 +868,71 @@ impl BpfManager {
...
+                if let Some(map) = self.maps.get_mut(&id) {
+                    map.used_by.push(id);
```
Comment (Billy99) [inline bpfd/src/bpf.rs:882]: Except here. Need to add this program's `id` to the map owner's used_by list.

File: bpfd/src/bpf.rs [hunk 5]
```diff
@@ -875,66 +868,71 @@ impl BpfManager {
...
-                .programs
-                .get_mut(&id)
-                .expect("Program should be loaded");
-
-            let map = BpfMap { used_by: vec![id] };
-
-            self.maps.insert(id, map);
+        match map_owner_id {
+            Some(id) => {
+                if let Some(map) = self.maps.get_mut(&id) {
```
Comment (Billy99) [inline bpfd/src/bpf.rs:881]:
```suggestion
                if let Some(map) = self.maps.get_mut(&owner_id) {
```

File: bpfd/src/bpf.rs [hunk 6]
```diff
@@ -875,66 +868,71 @@ impl BpfManager {
...
+                    map.used_by.push(id);
+
+                    // If map owner program still exists update it's maps_used_by_field
+                    if let Some(program) = self.programs.get_mut(&id) {
```
Comment (Billy99) [inline bpfd/src/bpf.rs:885]: Now `program.data.used_by` is what is returned in list/get, so this `if` block needs to be in a loop, and all programs with an instance in this `map.used_by` need to be updated with this new `id`.

The same will need to be done when a program in the `map.used_by` is unloaded. Also, `program.data.used_by` is not being updated on a bpfd restart.

The problem is you are trying to keep the same data in two different places and now you have to keep them in sync. And its a many `program.data.used_by` to one `map.used_by`. It's probably best to not keep `used_by` in ProgramData.

File: bpfctl/src/main.rs [hunk 7]
```diff
@@ -88,12 +92,7 @@ struct LoadFileArgs {
 
     /// Required: Name of the ELF section from the object file.
     #[clap(short, long)]
-    section_name: String,
-
-    /// Optional: Program uuid to be used by bpfd. If not specified, bpfd will generate
-    /// a uuid.
-    #[clap(long, verbatim_doc_comment)]
-    id: Option<String>,
+    name: String,
```
[the remark this answers is not in the document]
Comment (astoycos) [reply bpfctl/src/main.rs:95]: Nope, literally just the name

**Pull request merged**